src/OptimalTransport.jl

```diff
@@ -672,4 +675,156 @@ function quadreg(mu, nu, C, ϵ; θ=0.1, tol=1e-5, maxiter=50, κ=0.5, δ=1e-5)
+
+A pre-computed optimal transport `plan` may be provided.
+"""
+function otcost(
+    c, μ::ContinuousUnivariateDistribution, ν::UnivariateDistribution; plan=nothing
+)
+    if plan === nothing
+        g(μ, ν, x) = c(quantile(μ, x), quantile(ν, x))
+        f(x) = g(μ, ν, x)
+        return quadgk(f, 0, 1)[1]
+    else
+        cost = quadgk(0, 1) do q
+            x = quantile(μ, q)
+            return c(x, plan(x))
+        end
+        return cost[1]
+    end
+end
+
+"""
+    otplan(c,mu::ContinuousUnivariateDistribution,nu::UnivariateDistribution)
+
+Compute the optimal transport plan between μ to ν, where
+both are 1-Dimensional distributions and the cost
+function is of the form ``c(x,y) = h(|x-y|)`` such that
+``h`` is a convex function.
+
+A pre-computed optimal transport `plan` may be provided.
+
+Return the optimal transport plan as a function
+  ```math
+  T(x)=F_\\nu^{-1}(F_\\mu(x))
+  ```
+where ``F_\\mu`` is the cdf function of ``\\mu``
+and ``F_\\nu^{-1}`` is the quantile function of ``\\nu``.
+"""
+function otplan(c, μ::ContinuousUnivariateDistribution, ν::UnivariateDistribution)
+    # Use T instead of γ to indicate that this is a Monge map.
+    T(x) = quantile(ν, cdf(μ, x))
+    return T
+end
+
+"""
+    otcost(c, μ::DiscreteNonParametric, ν::DiscreteNonParametric; plan=nothing)
+
+Calculates the optimal transport cost between μ to ν, where
+both are 1-Dimensional discrete distributions with finite support
+and the cost function is of the form ``c(x,y) = h(|x-y|)`` such that
+``h`` is a convex function.
+
+A pre-computed optimal transport `plan` may be provided.
+"""
+function otcost(c, μ::DiscreteNonParametric, ν::DiscreteNonParametric; plan=nothing)
+    if plan === nothing
+        return _ot_cost_plan(c, μ, ν; get=:cost)
+    else
+        return dot(StatsBase.pairwise(c, μ.support, ν.support), plan)
+    end
+
+    return cost
+end
+
+"""
+    otplan(c, μ::DiscreteNonParametric, ν::DiscreteNonParametric; plan=nothing)
+
+Calculates the optimal transport cost between μ to ν, where
+both are 1-Dimensional discrete distributions with finite support
+and the cost function is of the form ``c(x,y) = h(|x-y|)`` such that
+``h`` is a convex function.
+
+Return the optimal transport plan as a matrix.
+
+A pre-computed optimal transport `plan` may be provided.
+"""
+function otplan(c, μ::DiscreteNonParametric, ν::DiscreteNonParametric; plan=nothing)
+    return _ot_cost_plan(c, μ, ν; get=:plan)
+end
+
+"""
+    _ot_cost_plan(c, μ::DiscreteNonParametric, ν::DiscreteNonParametric; get=:plan)
+
+Compute the optimal transport cost and plan for
+μ an ν where both are 1-Dimensional discrete distributions with finite support
+and the cost function is of the form ``c(x,y) = h(|x-y|)`` such that
+``h`` is a convex function.
+
+Return the plan or the cost depending if `get=:plan` or `get=:cost`, respectively.
+"""
+function _ot_cost_plan(c, μ::DiscreteNonParametric, ν::DiscreteNonParametric; get=:plan)
```

Comment:
I don't think that's a good approach, the function is type unstable and we can just hope that constant propagation kicks in. I had in mind the following: https://gist.github.com/devmotion/6b987174bce2bedf18b074358be80198 Maybe just leave out the `DiscreteNonParametric` case and I open a separate PR after basic support for Distributions is merged?

Reply:
Sure, I'll just submit the other part.
If I may, I'd like to understand some of the choices you made regarding the code in the gist. I see that you kept the `_ot_cost_plan` function with the `get`. Is this just for show or you intend to use it?
Secondly, why do you create this `Base.iterate` function.
I do not quite understand the issue with the function. I mean, what do you mean by "we can just hope that constant propagation kicks in"? How could this code go wrong?
Thanks

Reply:
The `_ot_cost_plan` I just included as a comparison and to check that the other implementation gives the same results.

The `Base.iterate` function (and also `Base.IteratorSize` and `Base.IteratorEltype`) belong to the iteration interface in Julia: https://docs.julialang.org/en/v1/base/collections/#lib-collections-iteration You have to define them if you implement an iterator. The custom iterator in the gist will just iteratively return tuples `(i, j, w)` where `i` and `j` are the indices of the source and target and `w` is the flow between them. This can be used to compute the cost and the plan completely independently (one could also compute both at the same time).

It is problematic if the return type of a function can't be computed from the types of the inputs. This is so-called type instability. Since the compiler does not know the type of the result, it can't create optimized compiled code and instead has to fall back on dynamically operating with the result at runtime (which typically is significantly slower). In this function the return type depends on the value of `get` and therefore it can't be inferred from the input types.

If you hardcode literals such as `1` or `:cost` then sometimes the compiler is smart enough and is able to infer the return type nevertheless. This is so-called constant propagation. Even though constant propagation improved a lot in recent Julia versions there are no hard guarantees and it might work or not.

BTW a common way to deal with type-unstable functions (sometimes it is not possible to avoid type instabilities) are so-called function barriers: if the result whose type can't be inferred is passed on to another function then the compiler still optimizes the called function for the type that it was called with, so the type instability does not impact the performance of subsequent computations. However, to achieve this you have to call another function and therefore it's called function barrier.

Reply:
Sorry for taking so long to submit the PR. As asked, I removed the `DiscreteNonParametric` implementation. I've left the tests commented out, so you can just uncomment when submitting your PR. I'm interested to see how you'll implement this. I've read your gist, but didn't quite understand all the code, and the implementation from `POT.py` returns both the cost and the transport plan.